controllers/oscmachine_keypair_controller.go

```diff
@@ -0,0 +1,121 @@
+package controllers
+
+import (
+	"context"
+	"fmt"
+
+	infrastructurev1beta1 "github.com/outscale-dev/cluster-api-provider-outscale.git/api/v1beta1"
+	"github.com/outscale-dev/cluster-api-provider-outscale.git/cloud/scope"
+	"github.com/outscale-dev/cluster-api-provider-outscale.git/cloud/services/compute"
+	"github.com/outscale-dev/cluster-api-provider-outscale.git/cloud/services/security"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+func checkKeypairFormatParameters(machineScope *scope.MachineScope) (string, error) {
+	machineScope.Info("Check Keypair parameters")
+
+	var keypairSpec *infrastructurev1beta1.OscKeypair
+	nodeSpec := machineScope.GetNode()
+	if nodeSpec.KeyPair.Name == "" {
+		nodeSpec.SetKeyPairDefaultValue()
+		keypairSpec = &nodeSpec.KeyPair
+	} else {
+		keypairSpec = machineScope.GetKeypair()
+	}
+
+	keypairName := keypairSpec.keypairName
+	keypairTagName, err := compute.ValidateKeypairName(keypairName)
+	if err != nil {
+		return keypairTagName, err
+	}
+
+	publicKey := keypairSpec.PublicKey
+	_, err = security.ValidateKeypairPublic(publicKey)
+	if err != nil {
+		return keypairTagName, err
+	}
+	return "", nil
+}
+
+// getKeyPairResourceId return the keypairName from the resourceMap base on resourceName (tag name + cluster uid)
+func getKeyPairResourceId(resourceName string, machineScope *scope.MachineScope) (string, error) {
+	keypairRef := machineScope.GetKeypairRef()
+	if keypairName, ok := keypairRef.ResourceMap[resourceName]; ok {
+		return keypairName, nil
+	} else {
+		return "", fmt.Errorf("%s does not exist", resourceName)
+	}
+}
+
+// reconcileKeypair reconcile the keypair of the machine
+func reconcileKeypair(ctx context.Context, machineScope *scope.MachineScope, keypairSvc security.OscKeyPairInterface) (reconcile.Result, error) {
+	machineScope.Info("Create Keypair or add existing one")
+	var keypairSpec *infrastructurev1beta1.OscKeypair
+	keypairSpec = machineScope.GetKeypair()
+	keypairRef := machineScope.GetKeypairRef()
+	keypairName := keypairSpec.Name + "-" + machineScope.GetUID()
+
+	if len(keypairRef.ResourceMap) == 0 {
+		keypairRef.ResourceMap = make(map[string]string)
+	}
+	if keypairSpec.ResourceId != "" {
+		keypairRef.ResourceMap[keypairName] = keypairSpec.ResourceId
+	}
+
+	keypair, err := keypairSvc.GetKeyPair(keypairName)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+	if keypair == nil || keypairSpec.ResourceId == "" {
+		goto CREATE
+	}
+
+	if keypairSpec.PublicKey != "" {
```

Comment:
Maybe before create ?

Reply:
it will check for the publickey of the key pair created or the existing one,